app/lib/actions/katello/host/update_system_purpose.rb

```diff
@@ -0,0 +1,27 @@
+module Actions
+  module Katello
+    module Host
+      class UpdateSystemPurpose < Actions::EntryAction
+        def plan(host, service_level, purpose_role, purpose_usage, purpose_addons)
+          fail _("Host %s has not been registered with subscription-manager.") % host.name unless host.subscription_facet
+          purpose_addon_objects = ::Katello::PurposeAddon.where(name: purpose_addons).first_or_create! unless purpose_addons.nil?
+
+          host.subscription_facet.service_level = service_level unless service_level.nil?
+          host.subscription_facet.purpose_role = purpose_role unless purpose_role.nil?
+          host.subscription_facet.purpose_usage = purpose_usage unless purpose_usage.nil?
+          host.subscription_facet.purpose_addons.clear if purpose_addons == [] && !purpose_addons.nil?
```

Comment:
I'm seeing a fair bit of conditionals related to the addons which sparked my interest. The more logic in the code -> the more execution paths thru the code -> the harder the code is to understand -> the harder the code is  to maintain -> the harder the code is to test!

Let's think about the requirements:
- if the input `purpose_addons` is nil, preserve whatever purpose addons are there on the facet. ie do nothing
- if the input `purpose addons` is an empty array, clear the purpose addons
- if the input  `purpose_addons` is an array with some items, replace the current purpose addons on the facet entirely

```

if purpose_addons
  if purpose_addons.empty?
    host.subscription_facet.purpose_addons = [] # i don't think we should use clear if we can avoid - it runs an extra query, but not sure if active record is smart enough to understand the empty array here
  else
    katello_purpose_addons = ::Katello::PurposeAddon.where(name: purpose_addons).first_or_create!
    host.subscription_facet.purpose_addons = katello_purpose_addons
  end
end
```

I'm not asking you to copy and paste what I wrote here, but feel free to use it or do something like it. I mainly wanted to illustrate how easy we can make the code to understand by keeping related things to their own block, if possible.